File: repos/openrappter/typescript/src/gateway/streaming.ts
/**
 * Enhanced streaming support for agent responses.
 *
 * StreamBlock is the atomic unit of streamed output. Each block has a type
 * ('text', 'tool_call', 'tool_result', 'thinking', 'error'), and may
 * accumulate content via deltas before being marked done.
 *
 * StreamManager maintains sessions keyed by a caller-supplied id, notifies
 * subscribers on every mutation, and integrates with the WebSocket gateway
 * for real-time delivery via broadcastEvent.
 */

import { randomUUID } from 'crypto';

// ── Types ─────────────────────────────────────────────────────────────────

export type BlockType = 'text' | 'tool_call' | 'tool_result' | 'thinking' | 'error';

export interface StreamBlock {
  id: string;
  type: BlockType;
  content: string;
  /** Incremental content appended in the most recent pushDelta call */
  delta?: string;
  metadata?: Record<string, unknown>;
  done: boolean;
  timestamp: number;
}

export type SessionStatus = 'active' | 'complete' | 'error';

export interface StreamSession {
  id: string;
  blocks: StreamBlock[];
  status: SessionStatus;
  createdAt: number;
  completedAt?: number;
}

/** Callback invoked whenever a block is pushed or updated */
export type BlockCallback = (block: StreamBlock, session: StreamSession) => void;

// ── Gateway integration shim ───────────────────────────────────────────────

/**
 * Minimal interface for the GatewayServer broadcastEvent method so that
 * StreamManager can push blocks over WebSocket without a hard dependency on
 * the full GatewayServer class.
 */
export interface GatewayBroadcaster {
  broadcastEvent(event: string, payload: unknown, filter?: (conn: unknown) => boolean): void;
}

// ── StreamManager ─────────────────────────────────────────────────────────

export class StreamManager {
  private sessions = new Map<string, StreamSession>();
  private subscribers = new Map<string, Set<BlockCallback>>();
  private gateway?: GatewayBroadcaster;

  constructor(gateway?: GatewayBroadcaster) {
    this.gateway = gateway;
  }

  /**
   * Attach (or replace) the gateway broadcaster used for real-time delivery.
   */
  setGateway(gateway: GatewayBroadcaster): void {
    this.gateway = gateway;
  }

  // ── Session lifecycle ──────────────────────────────────────────────────

  /**
   * Start a new streaming session. If a session with this id already exists
   * it is reset to an empty active state.
   */
  createSession(id: string): StreamSession {
    const session: StreamSession = {
      id,
      blocks: [],
      status: 'active',
      createdAt: Date.now(),
    };
    this.sessions.set(id, session);
    return session;
  }

  /**
   * Mark a session as complete. No further blocks should be pushed after this.
   */
  complete(sessionId: string): StreamSession | undefined {
    const session = this.sessions.get(sessionId);
    if (!session) return undefined;
    session.status = 'complete';
    session.completedAt = Date.now();
    return session;
  }

  /**
   * Mark a session as errored.
   */
  error(sessionId: string): StreamSession | undefined {
    const session = this.sessions.get(sessionId);
    if (!session) return undefined;
    session.status = 'error';
    session.completedAt = Date.now();
    return session;
  }

  /** Return the current session snapshot (blocks are a reference, not a copy). */
  getSession(sessionId: string): StreamSession | undefined {
    return this.sessions.get(sessionId);
  }

  // ── Block operations ───────────────────────────────────────────────────

  /**
   * Push a new block into the session. If a block with the same id already
   * exists it is replaced. Returns the (potentially updated) block.
   */
  pushBlock(sessionId: string, block: Omit<StreamBlock, 'id' | 'timestamp'> & { id?: string }): StreamBlock {
    const session = this.requireSession(sessionId);

    const resolved: StreamBlock = {
      id: block.id ?? randomUUID(),
      type: block.type,
      content: block.content,
      delta: block.delta,
      metadata: block.metadata,
      done: block.done,
      timestamp: Date.now(),
    };

    const existing = session.blocks.findIndex((b) => b.id === resolved.id);
    if (existing >= 0) {
      session.blocks[existing] = resolved;
    } else {
      session.blocks.push(resolved);
    }

    this.notify(sessionId, resolved, session);
    return resolved;
  }

  /**
   * Append a text delta to an existing block's content. Creates the block as
   * a 'text' block if it does not yet exist. Returns the updated block.
   */
  pushDelta(sessionId: string, blockId: string, delta: string): StreamBlock {
    const session = this.requireSession(sessionId);

    let block = session.blocks.find((b) => b.id === blockId);
    if (!block) {
      block = {
        id: blockId,
        type: 'text',
        content: '',
        done: false,
        timestamp: Date.now(),
      };
      session.blocks.push(block);
    }

    block.content += delta;
    block.delta = delta;
    block.timestamp = Date.now();

    this.notify(sessionId, block, session);
    return block;
  }

  // ── Subscriptions ──────────────────────────────────────────────────────

  /**
   * Subscribe to block updates for a session. The callback is invoked
   * synchronously on each pushBlock / pushDelta call.
   * Returns an unsubscribe function.
   */
  onBlock(sessionId: string, callback: BlockCallback): () => void {
    if (!this.subscribers.has(sessionId)) {
      this.subscribers.set(sessionId, new Set());
    }
    this.subscribers.get(sessionId)!.add(callback);

    return () => {
      const set = this.subscribers.get(sessionId);
      if (set) {
        set.delete(callback);
        if (set.size === 0) this.subscribers.delete(sessionId);
      }
    };
  }

  // ── Cleanup ────────────────────────────────────────────────────────────

  /** Remove a completed session and its subscribers from memory. */
  deleteSession(sessionId: string): void {
    this.sessions.delete(sessionId);
    this.subscribers.delete(sessionId);
  }

  /** Return the number of active (non-complete, non-error) sessions. */
  activeSessions(): number {
    let count = 0;
    for (const s of this.sessions.values()) {
      if (s.status === 'active') count++;
    }
    return count;
  }

  // ── Private ────────────────────────────────────────────────────────────

  private requireSession(sessionId: string): StreamSession {
    const session = this.sessions.get(sessionId);
    if (!session) throw new Error(`StreamSession '${sessionId}' not found`);
    return session;
  }

  private notify(sessionId: string, block: StreamBlock, session: StreamSession): void {
    // Notify local subscribers
    const set = this.subscribers.get(sessionId);
    if (set) {
      for (const cb of set) {
        try { cb(block, session); } catch { /* isolate subscriber errors */ }
      }
    }

    // Deliver over WebSocket gateway if attached
    if (this.gateway) {
      try {
        this.gateway.broadcastEvent('stream.block', {
          sessionId,
          block,
          sessionStatus: session.status,
        });
      } catch { /* ignore gateway errors */ }
    }
  }
}

/** Module-level singleton for convenience */
export const streamManager = new StreamManager();

File: repos/rappterbook-agent/python/openrappter/gateway/streaming.py
"""
Enhanced streaming support for agent responses.

StreamBlock is the atomic unit of streamed output. Each block has a type
('text', 'tool_call', 'tool_result', 'thinking', 'error'), and may
accumulate content via deltas before being marked done.

StreamManager maintains sessions keyed by a caller-supplied id, notifies
subscribers on every mutation, and mirrors the TypeScript StreamManager API.
"""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Literal, Optional

# ── Types ─────────────────────────────────────────────────────────────────

BlockType = Literal["text", "tool_call", "tool_result", "thinking", "error"]
SessionStatus = Literal["active", "complete", "error"]

BlockCallback = Callable[["StreamBlock", "StreamSession"], None]


@dataclass
class StreamBlock:
    id: str
    type: BlockType
    content: str
    done: bool
    timestamp: float
    delta: Optional[str] = None
    metadata: Optional[Dict] = None


@dataclass
class StreamSession:
    id: str
    blocks: List[StreamBlock]
    status: SessionStatus
    created_at: float
    completed_at: Optional[float] = None


# ── StreamManager ─────────────────────────────────────────────────────────

class StreamManager:
    """Manages streaming sessions, blocks, deltas, and subscriber callbacks."""

    def __init__(self) -> None:
        self._sessions: Dict[str, StreamSession] = {}
        self._subscribers: Dict[str, List[BlockCallback]] = {}

    # ── Session lifecycle ──────────────────────────────────────────────────

    def create_session(self, session_id: Optional[str] = None) -> StreamSession:
        """Start a new streaming session. Generates a UUID if no id supplied."""
        sid = session_id if session_id is not None else str(uuid.uuid4())
        session = StreamSession(
            id=sid,
            blocks=[],
            status="active",
            created_at=time.time(),
        )
        self._sessions[sid] = session
        return session

    def complete(self, session_id: str) -> Optional[StreamSession]:
        """Mark a session as complete with a completion timestamp."""
        session = self._sessions.get(session_id)
        if session is None:
            return None
        session.status = "complete"
        session.completed_at = time.time()
        return session

    def error(self, session_id: str) -> Optional[StreamSession]:
        """Mark a session as errored with a completion timestamp."""
        session = self._sessions.get(session_id)
        if session is None:
            return None
        session.status = "error"
        session.completed_at = time.time()
        return session

    def get_session(self, session_id: str) -> Optional[StreamSession]:
        """Return the current session snapshot or None if not found."""
        return self._sessions.get(session_id)

    # ── Block operations ───────────────────────────────────────────────────

    def push_block(
        self,
        session_id: str,
        block_type: BlockType,
        content: str,
        metadata: Optional[Dict] = None,
    ) -> StreamBlock:
        """Create a new block, append it to the session, and notify subscribers."""
        session = self._require_session(session_id)
        block = StreamBlock(
            id=str(uuid.uuid4()),
            type=block_type,
            content=content,
            done=False,
            timestamp=time.time(),
            metadata=metadata,
        )
        session.blocks.append(block)
        self._notify(session_id, block)
        return block

    def push_delta(self, session_id: str, block_id: str, delta: str) -> StreamBlock:
        """Append a delta to an existing block's content.

        If no block with block_id exists in the session, a new 'text' block is
        created and the delta becomes its initial content.
        """
        session = self._require_session(session_id)
        block = next((b for b in session.blocks if b.id == block_id), None)
        if block is None:
            block = StreamBlock(
                id=block_id,
                type="text",
                content="",
                done=False,
                timestamp=time.time(),
            )
            session.blocks.append(block)
        block.content += delta
        block.delta = delta
        block.timestamp = time.time()
        self._notify(session_id, block)
        return block

    # ── Subscriptions ──────────────────────────────────────────────────────

    def on_block(self, session_id: str, callback: BlockCallback) -> Callable[[], None]:
        """Register a callback invoked on every pushBlock / pushDelta call.

        Returns an unsubscribe function that removes the callback.
        """
        if session_id not in self._subscribers:
            self._subscribers[session_id] = []
        self._subscribers[session_id].append(callback)

        def unsubscribe() -> None:
            subs = self._subscribers.get(session_id)
            if subs is not None and callback in subs:
                subs.remove(callback)
                if not subs:
                    del self._subscribers[session_id]

        return unsubscribe

    # ── Computed properties ────────────────────────────────────────────────

    @property
    def active_sessions(self) -> int:
        """Return the count of sessions with status 'active'."""
        return sum(1 for s in self._sessions.values() if s.status == "active")

    # ── Private ────────────────────────────────────────────────────────────

    def _require_session(self, session_id: str) -> StreamSession:
        session = self._sessions.get(session_id)
        if session is None:
            raise KeyError(f"StreamSession '{session_id}' not found")
        return session

    def _notify(self, session_id: str, block: StreamBlock) -> None:
        session = self._sessions.get(session_id)
        if session is None:
            return
        for cb in list(self._subscribers.get(session_id, [])):
            try:
                cb(block, session)
            except Exception:
                pass  # isolate subscriber errors


# Module-level singleton for convenience
stream_manager = StreamManager()
